lib/bricks-net-api-types.js

```diff
@@ -0,0 +1,1176 @@
+						body_length = parseInt(value, 10);
+					}
+					else if (key === kTransferEncodingHeaderKey) {
+						if (value === kTransferEncodingChunkedValue) {
+							chunked_transfer_encoding = true;
+						}
+					}
+				}
+			}
+			else {
+				_this._DEBUG_LOG('TemplatedHTTPRequestData: part5 else');
+				
+				if (!chunked_transfer_encoding) {
+					_this._DEBUG_LOG('TemplatedHTTPRequestData: part5 else !chunked_transfer_encoding');
+					
+					body_offset = next_line_offset;
+					if (body_length !== UINT_MAX) {
+						length_cap = body_offset + body_length;
+						if (length_cap + 1 > _this.buffer_.length) {
+							var delta_size = length_cap + 1 - _this.buffer_.length;
+							var buffer_new_length = Math.min(delta_size, buffer_max_growth_due_to_content_length);
+							_this.buffer_ = Buffer.concat([
+								_this.buffer_, 
+								new Buffer(buffer_new_length - _this.buffer_.length)
+							], buffer_new_length);
+						}
+					}
+					else {
+						length_cap = body_offset;
+					}
+				}
+				else {
+					_this._DEBUG_LOG('TemplatedHTTPRequestData: part5 else -> receiving_body_in_chunks');
+					
+					receiving_body_in_chunks = true;
+				}
+			}
+			current_line_offset = next_line_offset;
+			
+			_this._DEBUG_LOG('TemplatedHTTPRequestData: part5 continue while2',
+				'current_line_offset ===', current_line_offset);
+			
+			// continue while3
+			part4();
+		}
+	})();
+}
+/* istanbul ignore next: unused method */
+TemplatedHTTPRequestData.prototype.Method = function () {
+	return this.method_;
+};
+/* istanbul ignore next: unused method */
+TemplatedHTTPRequestData.prototype.URL = function () {
+	return this.url_;
+};
+TemplatedHTTPRequestData.prototype.RawPath = function () {
+	return this.raw_path_;
+};
+TemplatedHTTPRequestData.prototype.HasBody = function () {
+	return (this.body_buffer_begin_ !== null || this.chunked_body_buffer_ !== null);
+};
+TemplatedHTTPRequestData.prototype.Body = function () {
+	var _this = this;
+	if (!_this.prepared_body_) {
+		if (_this.body_buffer_begin_ !== null) {
+			_this.prepared_body_ = _this.buffer_.toString('utf8', _this.body_buffer_begin_, _this.body_buffer_end_);
+		}
+		else if (_this.chunked_body_buffer_ !== null) {
+			_this.prepared_body_ = _this.chunked_body_buffer_;
+		}
+		else {
+			BRICKS_THROW(HTTPNoBodyProvidedException());
+		}
+	}
+	return _this.prepared_body_;
+};
+TemplatedHTTPRequestData.prototype.BodyBegin = notImplementedMethod();
+TemplatedHTTPRequestData.prototype.BodyEnd = notImplementedMethod();
+/* istanbul ignore next: unused method */
+TemplatedHTTPRequestData.prototype.BodyLength = function () {
+	var _this = this;
+	if (_this.body_buffer_begin_ !== null) {
+		return (_this.body_buffer_end_ - _this.body_buffer_begin_);
+	}
+	else if (_this.chunked_body_buffer_ !== null) {
+		return _this.chunked_body_buffer_.length;
+	}
+	else {
+		BRICKS_THROW(HTTPNoBodyProvidedException());
+	}
+};
+TemplatedHTTPRequestData.prototype.OnHeader = virtualMethod();
```

Comment:
Sure, it should be `makeVirtualMethod` (in this code base I use 'makeSomething').

In general I prefer not to rewrite C++ code into JavaScript, so I prefer to have no virtual methods.

If absolutely required, I've got something like 
`X.prototype.method = utils.declareNotImplementedMethod('X#method');`
and
`utils.declareNotImplementedMethod = function (methodName) { return function () { throw new NotImplementedError(methodName); } };`
in another project's code base.